src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -173,56 +177,118 @@ private static void setLandingPageTitle(String filePath, String pageTitle) throw
-        RepoCloner repoCloner = new RepoCloner();
-        RepoLocation clonedRepoLocation = null;
-
         List<RepoLocation> repoLocationList = new ArrayList<>(repoLocationMap.keySet());
 
-        RepoLocation currRepoLocation = repoLocationList.get(0);
-        repoCloner.cloneBare(repoLocationMap.get(currRepoLocation).get(0));
+        // Fixed thread pools are used to limit the number of threads used by cloning and analysis jobs at any one time
+        ExecutorService cloneExecutor = Executors.newFixedThreadPool(numCloningThreads);
+        ExecutorService analyzeExecutor = Executors.newFixedThreadPool(numAnalysisThreads);
 
-        List<Path> generatedFiles = new ArrayList<>();
-        for (int index = 1; index <= repoLocationList.size(); index++) {
-            RepoLocation nextRepoLocation = (index < repoLocationList.size()) ? repoLocationList.get(index) : null;
-            clonedRepoLocation = repoCloner.getClonedRepoLocation();
+        List<CompletableFuture<AnalyzeJobOutput>> analyzeJobFutures = new ArrayList<>();
+        for (RepoLocation location : repoLocationList) {
+            List<RepoConfiguration> configsToAnalyze = repoLocationMap.get(location);
 
-            // Clones the next location while analyzing the previously cloned repos in parallel.
-            if (nextRepoLocation != null) {
-                repoCloner.cloneBare(repoLocationMap.get(nextRepoLocation).get(0));
-            }
+            // The `CompletableFuture.supplyAsync` method is used to clone the repo in parallel.
+            // Note that the `cloneExecutor` is passed as a parameter to ensure that the number of threads used
+            // for cloning is no more than `numCloningThreads`.
+            CompletableFuture<CloneJobOutput> cloneFuture = CompletableFuture.supplyAsync(() ->
+                    cloneRepo(configsToAnalyze.get(0), location), cloneExecutor);
 
-            if (clonedRepoLocation == null) {
-                handleCloningFailed(configs, currRepoLocation);
-            } else {
-                generatedFiles.addAll(analyzeRepos(outputPath, configs, repoLocationMap.get(clonedRepoLocation),
-                        repoCloner.getCurrentRepoDefaultBranch()));
-            }
-            currRepoLocation = nextRepoLocation;
+            // The `thenApplyAsync` method is used to analyze the cloned repo in parallel.
+            // This ensures that the analysis job for each repo will only be run after the repo has been cloned.
+            // Note that the `analyzeExecutor` is passed as a parameter to ensure that the number of threads used
+            // for analysis is no more than `numAnalysisThreads`.
+            CompletableFuture<AnalyzeJobOutput> analyzeFuture = cloneFuture.thenApplyAsync(
+                    cloneJobOutput -> analyzeRepos(outputPath, configsToAnalyze, cloneJobOutput), analyzeExecutor);
+
+            analyzeJobFutures.add(analyzeFuture);
         }
+
```

Comment:
FYI, I changed this portion a bit to make it more readable. Instead of the streams, I switched bacj to a for loop. I feel that it makes the 2 lines relating to the `CompletableFutures` a little easier to parse.